NMS-12731: ResourceDao performance enhancements for the Time Series Storage API - Part 1

patrick-schweizer opened the pull request:
This PR contains performance improvements for the Timeseries Integration Layer. The overall ticket is: [NMS-12731: Optimize Performance of Timeseries Integration Layer](https://issues.opennms.org/browse/NMS-12731)

Since these are large changes I split them in multiple PRs. This PR addresses the following sub tasks:
* [NMS-12744](https://issues.opennms.org/browse/NMS-12744): Limit number of calls to find metrics by
  * querying for all metrics under a given resource path (instead of querying for only the next level).
  * caching the results by the index tags
* [NMS-12745](https://issues.opennms.org/browse/NMS-12745): Limit the calls to the timeseries_metatable:  by filling the cache already during writing (before it was only during reading)
* [NMS-12746](https://issues.opennms.org/browse/NMS-12746): Remove (if possible) the conversion from Opennms -> Newts -> TS

It also removes some unused code.

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-12731

patrick-schweizer commented:
@j-white I have here a first implementation that cuts down on the calls to TimeseriesStorage by querying for all metrics under a path (currently node level).
What do you think? How do we test if the performance is sufficient?

fooker commented on features/timeseries/src/main/java/org/opennms/netmgt/timeseries/meta/TimeSeriesMetaDataDao.java:
```diff
@@ -69,34 +69,35 @@
-                    .map(entry -> entry.get(meta.getName()))
-                    .isPresent()) {
-                uncached.add(meta);
+            Map<String, String> attributesForResource = cache.get(meta.getResourceId(), HashMap::new);
+            if(attributesForResource.get(meta.getName()) == null) {
```
```suggestion
            Map<String, String> attributesForResource = cache.getIfPresent(meta.getResourceId());
            if(attributesForResource != null && attributesForResource.contains(meta.getName())) {
```

To save the allocation

patrick-schweizer replied:
:+1:

fooker commented on opennms-model/src/main/java/org/opennms/netmgt/model/RrdGraphAttribute.java:
```diff
@@ -136,4 +137,19 @@ public String toString() {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        RrdGraphAttribute that = (RrdGraphAttribute) o;
+        return Objects.equals(m_name, that.m_name) &&
+                Objects.equals(m_relativePath, that.m_relativePath) &&
+                Objects.equals(m_rrdFile, that.m_rrdFile) &&
+                Objects.equals(m_resource, that.m_resource);
```
If you call the other side `that` you should prefix the instance with `this.` :-)

```suggestion
        return Objects.equals(this.m_name, that.m_name) &&
                Objects.equals(this.m_relativePath, that.m_relativePath) &&
                Objects.equals(this.m_rrdFile, that.m_rrdFile) &&
                Objects.equals(this.m_resource, that.m_resource);
```

patrick-schweizer replied:
:+1:

fooker commented on opennms-model/src/main/java/org/opennms/netmgt/model/StringPropertyAttribute.java:
```diff
@@ -89,4 +91,19 @@ public OnmsResource getResource() {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        StringPropertyAttribute that = (StringPropertyAttribute) o;
+        return Objects.equals(m_name, that.m_name) &&
+                Objects.equals(m_value, that.m_value) &&
+                Objects.equals(m_resource, that.m_resource);
```
```suggestion
        return Objects.equals(this.m_name, that.m_name) &&
                Objects.equals(this.m_value, that.m_value) &&
                Objects.equals(this.m_resource, that.m_resource);
```

patrick-schweizer commented on core/cache/src/main/java/org/opennms/core/cache/Cache.java:
```diff
@@ -84,6 +85,18 @@ public V get(K key) throws ExecutionException {
+        try {
+            return valueLoader.call();
+        } catch (Throwable t) {
+            throw new RuntimeException(t);
```
java.util.concurrent.Callable declares to throw `Exception`:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Callable.html
I assume that is the reason for the existing code.

fooker replied:
You are totally right. I mixed it up with the `Producer` interface. But this `Callable` is more appropriate as an error can happen and should be expected.

patrick-schweizer commented:
> A lot of calls from `Resource.gtId` hav been changed to `Metric.getFirstTagByKey` which _sounds_ to be less performent. Any insights?

That is true and due to the fact that we don't operate any more on the Newts objects. Before we did:
Opennms objects -> newts objects -> timeseries integration layer objects and revers
now we do:
Opennms objects -> timeseries integration layer objects and revers
(I removed the transformations from and to Newts objects, they were left over from the old module).

Having said that why do we have now this weird Metric.getFirstTagByKey? When designing the Integration layer we had those design principles in mind:
* very slim interface
* now knowledge of Opennms internals
* following the http://metrics20.org/spec/ as close as reasonable

This results in a metric not knowing anything about resourceId but just about Tags, thus the weird lookup.

I think the overhead is not significant enough to justify 'polluting' the interface with Opennms specifics but that is a trade off to discuss.

fooker commented:
> I think the overhead is not significant enough to justify 'polluting' the interface with Opennms specifics but that is a trade off to discuss.

Ok, this is fine. If there is a performance problem with this, we can tune later on.

patrick-schweizer commented:
@j-white not sure if you want to take a look as well.
This PR is pretty big, It contains all subtickets from: https://issues.opennms.org/browse/NMS-12731
It was reviewed by @fooker and is ready to be merged.

j-white commented:
Something isn't quite right with the resource caching. When trying to load the resource page for a node, the Cortex plugin logs the following:
```
16:07:03.669 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.695 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.718 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.733 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.753 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.770 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.793 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.812 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.839 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.858 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.878 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.898 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.924 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.942 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.960 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.976 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
16:07:03.999 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
```

patrick-schweizer commented:
> Something isn't quite right with the resource caching. When trying to load the resource page for a node, the Cortex plugin logs the following:
> 
> ```
> 16:07:03.669 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.695 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.718 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.733 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.753 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.770 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.793 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.812 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.839 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.858 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.878 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.898 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.924 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.942 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.960 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.976 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> 16:07:03.999 INFO [qtp1494965893-443] Retrieving metrics for tags: [_idx2w=(snmp:4,*)]
> ```

This is embarrassing :-o Maybe too much refactoring...
I fixed it and added a test: https://github.com/OpenNMS/opennms/pull/3025/commits/e37d635f56308d665565b64bf425edacd2d09f6e

j-white commented:
That fixed it. The performance is much better and the Cortex plugin is functional.
